corehq/apps/smsforms/tasks.py

```diff
@@ -125,8 +127,14 @@ def handle_due_survey_action(domain, contact_id, session_id):
             session.move_to_next_action()
             session.save()
         else:
-            # Close the session
-            session.close()
+            try:
+                session.close()
+            except TouchformsError as e:
+                try:
+                    self.retry(exc=e)
+                except MaxRetriesExceededError:
```

Comment:
I think the `exc` param lets you override the default value of the error thrown when max retries are exceeded, and that `MaxRetriesExceededError` is the default. So I think these two lines might conflict with each other semantically. See the param descriptions in https://docs.celeryproject.org/en/stable/reference/celery.app.task.html#celery.app.task.Task.retry for closer detail (I might not have it 100% right).

Reply:
Also I'm not sure if retrying this task is the right thing to do—it'll re-execute everything above this, which might not be right?

To make it crystal clear and also possibly something you can reuse in other places that call session.close(), perhaps you can add a celery task called `close_session` that does just this (and retries like this).